HW1/P6/P6.py

```diff
@@ -0,0 +1,219 @@
+    """
+    # Just one or more letters
+    a_word = re.compile(r"[\w']+")
+    # Match all the words in a given line
+    matches = a_word.findall(line)
+    
+    # findall returns a list, so...
+    return matches
+
+def filter_words(word):
+    """Function that uses regular expressions to filter out words that are made of
+    entirely capital letters, or entirely capital letters followed by a period, as per
+    the problem statement.
+    Note that the way we matched words, \w+, has already filtered out words made of only numbers.
+    """
+    
+    length_of_word = len(word)
+    
+    # Matches one or more capital letters in a row.
+    only_capitals = re.compile(r"[A-Z]{" + str(length_of_word) + "}")
+    
+    # Matches one or more capital letters in a row followed by a period.
+    only_capitals_period = re.compile(r"[A-Z]{" + str(length_of_word-1) + "}\.+")
+    
+    # Only numbers
+    only_numbers = re.compile(r"\d{" + str(length_of_word) + "}")
+    
+    # if it's a capital...
+    caps = only_capitals.match(word)
+    
+    # if it's a capital followed by a period...
+    cap_per = only_capitals_period.match(word)
+    
+    # if it's a number...
+    numbs = only_numbers.match(word)
+        
+    result = True
+    if (caps) or (cap_per) or (numbs):
+        result = False
+        
+    return result
+
+def count_occurrences(words):
+    """ Takes an iterable, words, and returns a list
+    of tuples of each word that appears in it and the number of times it appears.
+    To be used with map after groupByKey.
+    """
+    
+    results = []
+    dist_words = set(words)
+    orig_words = list(words)
+    
+    for word in dist_words:
+        c = orig_words.count(word)
+        results.append((word, c))
+        
+    return results
+
+def generate_phrase(markov, num_words):
+    """Takes a markov chain, assumed to be an RDD of the form described in the problem
+    and returns a randomly generated phrase with num_words words. The phrase is generated by
+    choosing a pair of words randomly from the RDD, weighting the next word by the relative
+    frequencies of occurrences of other words after that pair in Shakespeare, and then
+    continues to do so until we have num_words.
+    """
+    # Our counter
+    curr_num_words = 0
+    
+    # Start us off with a blank phrase
+    phrase = []
+
+    # First flag is for with replacement... doesn't matter because we just do one at a time
+    # Note that takeSample returns a list, so we just really want one element from it
+    next_three = markov.takeSample(True, 1)[0]
+    
+    # Get the first two words
+    word_1 = next_three[0][0]
+    word_2 = next_three[0][1]
+    
+    # Now get the list of possible words
+    poss_words = next_three[1]
+    
+    # And choose accordingly...
+    word_3 = weighted_choice(poss_words)
+    
+    # Add our words to the phrase...
+    phrase.append(word_1)
+    phrase.append(word_2)
+    phrase.append(word_3)
+    
```

Comment:
Why not just skip this part, set `next_pair` to a random pair of words, and let the loop run?